### ISSUE TITLE
postgres float vs double confusion with binary copy

### ISSUE THREAD

**Issue by dhalperi** [uwescience/myria#690]
fyi @ljorr1 @jortiz16 @domoritz 

1. Right now, Postgres uses 64-bit floats for both `FLOAT_TYPE` and `DOUBLE_TYPE`. See https://github.com/uwescience/myria/blob/master/src/edu/washington/escience/myria/accessmethod/JdbcAccessMethod.java#L382

2. The Binary copy code writes 32-bit floats for `FLOAT_TYPE` and 64-bit floats for `DOUBLE_TYPE`. See https://github.com/uwescience/myria/blob/6573fca6ba809d6d3eaf76d7ba90d1c86dff1518/src/edu/washington/escience/myria/PostgresBinaryTupleWriter.java#L101

This combination of features results in the following error during binary copy of `FLOAT_TYPE` data:

```java
ERROR 2015-02-08 16:54:45,648 [Nonblocking query executor#10] JdbcAccessMethod - Error inserting batch via PostgreSQL COPY
edu.washington.escience.myria.DbException: java.io.IOException: Write to copy failed.
	at edu.washington.escience.myria.accessmethod.JdbcAccessMethod.postgresCopyInsert(JdbcAccessMethod.java:135)
	at edu.washington.escience.myria.accessmethod.JdbcAccessMethod.tupleBatchInsert(JdbcAccessMethod.java:153)
	at edu.washington.escience.myria.operator.DbInsert.consumeTuples(DbInsert.java:166)
	at edu.washington.escience.myria.operator.RootOperator.fetchNextReady(RootOperator.java:123)
	at edu.washington.escience.myria.operator.Operator.nextReady(Operator.java:335)
	at edu.washington.escience.myria.parallel.LocalFragment.executeActually(LocalFragment.java:458)
	at edu.washington.escience.myria.parallel.LocalFragment.access$400(LocalFragment.java:43)
	at edu.washington.escience.myria.parallel.LocalFragment$1.call(LocalFragment.java:195)
	at edu.washington.escience.myria.parallel.LocalFragment$1.call(LocalFragment.java:175)
	at edu.washington.escience.myria.util.concurrent.ExecutableExecutionFuture.call(ExecutableExecutionFuture.java:193)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at edu.washington.escience.myria.util.concurrent.RenamingThreadFactory$1.run(RenamingThreadFactory.java:33)
Caused by: java.io.IOException: Write to copy failed.
	at org.postgresql.copy.PGCopyOutputStream.write(PGCopyOutputStream.java:82)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.write(BufferedOutputStream.java:95)
	at java.io.DataOutputStream.writeInt(DataOutputStream.java:197)
	at java.io.DataOutputStream.writeFloat(DataOutputStream.java:242)
	at edu.washington.escience.myria.PostgresBinaryTupleWriter.writeTuples(PostgresBinaryTupleWriter.java:111)
	at edu.washington.escience.myria.accessmethod.JdbcAccessMethod.postgresCopyInsert(JdbcAccessMethod.java:125)
	... 13 more
Caused by: org.postgresql.util.PSQLException: ERROR: insufficient data left in message
  Where: COPY public:MyriaSysTemp:snapshot1309Hash2, line 1, column ESNRate
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2198)
	at org.postgresql.core.v3.QueryExecutorImpl.processCopyResults(QueryExecutorImpl.java:966)
	at org.postgresql.core.v3.QueryExecutorImpl.writeToCopy(QueryExecutorImpl.java:856)
	at org.postgresql.core.v3.CopyInImpl.writeToCopy(CopyInImpl.java:51)
	at org.postgresql.copy.PGCopyOutputStream.writeToCopy(PGCopyOutputStream.java:123)
	at org.postgresql.copy.PGCopyOutputStream.write(PGCopyOutputStream.java:80)
	... 19 more
```

because Postgres is expecting 64 bits but we only send them 32.

(This issue does not cause problems during execution because we fall back to the old-style JDBC batch insert method.)

I don't particularly care how this issue is resolved -- I think some other JDBC database we tried to support did not have both types of floats -- but I think that Laurel and Jen are probably suffering bad performance during ingest and writeback of the Nchilada and Tipsy data.

**Comment by dhalperi** [comment]
@domoritz any opinions here? @ljorr1 @jortiz16 ?

If not, I may decide that the right solution is to make PostgresBinaryTupleWriter use 64-bits always for floats.

**Comment by domoritz** [comment]
If pg uses 64 bit for both, then I don'e see a better way than giving it 64 bit.

**Comment by dhalperi** [comment]
As desc states, Postgres itself has support for both 32- and 64-bit floats, I think another JDBC backend didn't and this is maybe an oversight. We could instead switch PG to use 32-bit storage for FLOAT_TYPE, which might necessitate a few "alter table" statements but that shouldn't be an issue

**Comment by senderista** [comment]
Is there any reason not to just change `typeToDbmsType()` to emit "REAL" for `FLOAT_TYPE` and "DOUBLE PRECISION" for `DOUBLE_TYPE`? If, as @dhalperi said, there are no serious consequences to users of doing so, what's the downside?

**Comment by dhalperi** [comment]
Maybe not. The main thing is the followup work to make the production cluster accurate: we need to `ALTER` the schema for all existing tables with floats in them.

**Comment by senderista** [comment]
fixed in #726
